File: packages/lua/sorting.lua
local sorting = {}

function sorting.BubbleSort(arr, less)
    local n = #arr
    for i = 1, n - 1 do
        for j = 1, n - i do
            if less(arr[j+1], arr[j]) then
                arr[j], arr[j+1] = arr[j+1], arr[j]
            end
        end
    end
end

function sorting.SelectionSort(arr, less)
    local n = #arr
    for i = 1, n - 1 do
        local minIndex = i
        for j = i + 1, n do
            if less(arr[j], arr[minIndex]) then
                minIndex = j
            end
        end
        arr[i], arr[minIndex] = arr[minIndex], arr[i]
    end
end

function sorting.InsertionSort(arr, less)
    local n = #arr
    for i = 2, n do
        local key = arr[i]
        local j = i - 1
        while j >= 1 and less(key, arr[j]) do
            arr[j+1] = arr[j]
            j = j - 1
        end
        arr[j+1] = key
    end
end

local function heapify(arr, n, i, less)
    local largest = i
    local left = 2 * i
    local right = 2 * i + 1

    if left <= n and less(arr[largest], arr[left]) then
        largest = left
    end

    if right <= n and less(arr[largest], arr[right]) then
        largest = right
    end

    if largest ~= i then
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest, less)
    end
end

function sorting.HeapSort(arr, less)
    local n = #arr

    for i = math.floor(n / 2), 1, -1 do
        heapify(arr, n, i, less)
    end

    for i = n, 2, -1 do
        arr[1], arr[i] = arr[i], arr[1]
        heapify(arr, i - 1, 1, less)
    end
end

function sorting.QuickSort(arr, less)
    local function partition(low, high)
        local pivot = arr[high]
        local i = low - 1
        for j = low, high - 1 do
            if less(arr[j], pivot) then
                i = i + 1
                arr[i], arr[j] = arr[j], arr[i]
            end
        end
        arr[i + 1], arr[high] = arr[high], arr[i + 1]
        return i + 1
    end

    local function quickSortHelper(low, high)
        if low < high then
            local pi = partition(low, high)
            quickSortHelper(low, pi - 1)
            quickSortHelper(pi + 1, high)
        end
    end

    quickSortHelper(1, #arr)
end

function sorting.MergeSort(arr, less)
    local function merge(low, mid, high)
        local n1 = mid - low + 1
        local n2 = high - mid
        local L = {}
        local R = {}
        for i = 1, n1 do L[i] = arr[low + i - 1] end
        for j = 1, n2 do R[j] = arr[mid + j] end
        
        local i, j, k = 1, 1, low
        while i <= n1 and j <= n2 do
            if not less(R[j], L[i]) then
                arr[k] = L[i]
                i = i + 1
            else
                arr[k] = R[j]
                j = j + 1
            end
            k = k + 1
        end
        while i <= n1 do
            arr[k] = L[i]
            i = i + 1
            k = k + 1
        end
        while j <= n2 do
            arr[k] = R[j]
            j = j + 1
            k = k + 1
        end
    end

    local function mergeSortHelper(low, high)
        if low < high then
            local mid = math.floor((low + high) / 2)
            mergeSortHelper(low, mid)
            mergeSortHelper(mid + 1, high)
            merge(low, mid, high)
        end
    end

    mergeSortHelper(1, #arr)
end

return sorting

File: packages/cpp/Sorting.cpp
#include "Sorting.hpp"
#include <algorithm>

void Sorting::bubbleSort(std::vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n - 1; ++i) {
        for (int j = 0; j < n - i - 1; ++j) {
            if (arr[j] > arr[j + 1]) {
                std::swap(arr[j], arr[j + 1]);
            }
        }
    }
}

void Sorting::selectionSort(std::vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n - 1; ++i) {
        int min_idx = i;
        for (int j = i + 1; j < n; ++j) {
            if (arr[j] < arr[min_idx]) {
                min_idx = j;
            }
        }
        std::swap(arr[i], arr[min_idx]);
    }
}

void Sorting::insertionSort(std::vector<int>& arr) {
    int n = arr.size();
    for (int i = 1; i < n; ++i) {
        int key = arr[i];
        int j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j = j - 1;
        }
        arr[j + 1] = key;
    }
}

void Sorting::heapSort(std::vector<int>& arr) {
    auto heapify = [&](int n, int i, auto& heapify_ref) -> void {
        int largest = i;
        int l = 2 * i + 1;
        int r = 2 * i + 2;

        if (l < n && arr[l] > arr[largest]) {
            largest = l;
        }
        if (r < n && arr[r] > arr[largest]) {
            largest = r;
        }

        if (largest != i) {
            std::swap(arr[i], arr[largest]);
            heapify_ref(n, largest, heapify_ref);
        }
    };

    int n = arr.size();
    for (int i = n / 2 - 1; i >= 0; i--) {
        heapify(n, i, heapify);
    }

    for (int i = n - 1; i > 0; i--) {
        std::swap(arr[0], arr[i]);
        heapify(i, 0, heapify);
    }
}

static void quickSortHelper(std::vector<int>& arr, int low, int high) {
    if (low < high) {
        int pivot = arr[high];
        int i = low - 1;
        for (int j = low; j <= high - 1; j++) {
            if (arr[j] < pivot) {
                i++;
                std::swap(arr[i], arr[j]);
            }
        }
        std::swap(arr[i + 1], arr[high]);
        int pi = i + 1;
        quickSortHelper(arr, low, pi - 1);
        quickSortHelper(arr, pi + 1, high);
    }
}

void Sorting::quickSort(std::vector<int>& arr) {
    if (arr.empty()) return;
    quickSortHelper(arr, 0, arr.size() - 1);
}

static void merge(std::vector<int>& arr, int l, int m, int r) {
    int n1 = m - l + 1;
    int n2 = r - m;
    std::vector<int> L(n1), R(n2);
    for (int i = 0; i < n1; i++) L[i] = arr[l + i];
    for (int j = 0; j < n2; j++) R[j] = arr[m + 1 + j];
    int i = 0, j = 0, k = l;
    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) arr[k++] = L[i++];
        else arr[k++] = R[j++];
    }
    while (i < n1) arr[k++] = L[i++];
    while (j < n2) arr[k++] = R[j++];
}

static void mergeSortHelper(std::vector<int>& arr, int l, int r) {
    if (l >= r) return;
    int m = l + (r - l) / 2;
    mergeSortHelper(arr, l, m);
    mergeSortHelper(arr, m + 1, r);
    merge(arr, l, m, r);
}

void Sorting::mergeSort(std::vector<int>& arr) {
    if (arr.empty()) return;
    mergeSortHelper(arr, 0, arr.size() - 1);
}
